Adding a Versioning topic to MFTF (#3781)
magento/devdocs: mftf/2.3/versioning.md

## Backwards Compatibility Definition

Backwards compatibility is when a test undergoes changes, but achieves the same results as before and remains compatible with potential test customizations.

Different types of changes that are backwards compatible:

- **Test Flow change (Test/ActionGroup)** - A backwards compatible modification of a test flow would not diminish the original set of actions in the test.
Some modifications may change an action's sequence or behavior, but still allows an extension to achieve the same test results without changing the test extension.
- **Test Entity change (Data/Section/Page/Metadata)** - Adding new values or updating a `value` of existing entities which do **NOT** require updates to the test.
- **Test Annotation change** - Annotations can be changed and are considered backward compatible changes.
However, removing or changing a `<group />` annotation is considered a backward incompatible change.
- Changes which delete and/or rename (Test/Action Group/Data/Metadata/Page/Section/Action)'s id attribute are considered as a backward incompatible change. Changing a reference to a data entity is considered a backward incompatible change.

> 

Reviewer: Suggestion -- tie the backward compatibility section to the versioning schema instead of calling it out as a definition. That is,  the versioning schema is tied to the type of changes included in test updates to be packaged for release....

Also -- backwards ==> backward

For example, the versioning schema is tied to the types of changes included in a test update.  Then, mention two types of changes -- backward compatible with examples, and backward incompatible.

Author: This section discussed changes to tests rather than MFTF. Now that I know that distinction, this section belonged elsewhere and I removed it.

## Find your MFTF version number

To find the current version of MFTF you are using, run the Magento CLI command:

```bash
cd <magento_root>/
vendor/bin/mftf --version
```

## Versioning Policy

MFTF versioning policy follows guidelines from [Semantic Versioning](https://semver.org/).

### 3-component version numbers

    X.Y.Z
    | | |
    | | +-- Backward Compatible changes (Patch release - bug fixes)
    | +---- Backward Compatible changes (Minor release - new tests)
    +------ Backward Incompatible changes (Major release)

> 

Reviewer: It might be clearer if you begin with an example version number, and then show how it changes for each type of release.  For example, start with version 1.0.0 ---   version after patch release -->  1.0.1 --> version after minor release  1.1.0 --> version after major release --> 2.0.0

Author: Done.

### Z release - patch

  Patch version **Z** MUST be incremented if only backward compatible changes to tests are introduced.

>   For example, a fix which aims to resolve test flakiness might:

Author: Reworked the sections here. Thanks. I think 'incremented' is the proper verb here. The language, including the MUST comes from https://semver.org/

- Update an unreliable selector.
- Add a wait for an element.
- Update a data entity value.
  
### Y release - minor

  Minor version **Y** MUST be incremented if a new, backwards compatible test or test entity is introduced.
  It MUST be incremented if any test or test entity is marked as deprecated.
  It MAY include patch level changes. Patch version MUST be reset to 0 when minor version is incremented.

### X release - major

>   Major version **X** MUST be incremented if any backwards incompatible changes are introduced to a test or test entity.

Reviewer: A major release introduces backward incompatible changes to a test or test entity. ....

Move the version number change after the example.

Author: Thanks.

>   It MAY include minor and patch level changes. Patch and minor version MUST be reset to 0 when major version is incremented.

Reviewer: ```suggestion
  It can include minor and patch level changes. You must reset the patch and minor version to 0 when you change the major version.
```

## Incompatible entity attribute changes

> Changing any of the following attributes will cause a backward incompatible change.

Reviewer: ```suggestion
Changing any of the following entity attributes or actions introduces a backward incompatible change.
```

X.Y.Z
| | |
| | +-- Backward Compatible changes (Patch release - bug fixes, small additions)
| +---- Backward Compatible changes (Minor release - small new features, bug fixes)
+------ Backward Incompatible changes (Major release - new features and/or major changes)

For example:

- Magento 2 ships with MFTF 2.3.9
- A patch is added to fix a bug: 2.3.10 (Increment Z = backward compatible change)
- New action command added: 2.4.0 (Increment Y, set Z to 0 = backward compatible change)
- New action added: 2.4.1 (Increment Z = backward compatible change)
- Major new features added to MFTF to support changes in Magento codebase: 3.0.0. (Increment X, reset Y and Z to 0 = backward incompatible change)

### Z release - patch

Patch version **Z** MUST be incremented if only backward compatible changes are introduced.
  
### Y release - minor

Minor version **Y** MUST be incremented if new, backward compatible features are introduced.
It MUST be incremented if any test or test entity is marked as deprecated.
It MAY include patch level changes. Patch version MUST be reset to 0 when minor version is incremented.

### X release - major

> A major release introduces backward incompatible changes. This MUST increment the major version number.

Reviewer: ```suggestion
Major version **X** MUST be incremented for a release that introduces backward incompatible changes.
```

Reviewer: Suggest making this change to be consistent with description in other sections.

> It can include minor and patch level changes. You must reset the patch and minor version to 0 when you change the major version.

Reviewer: ```suggestion
A major release can also include minor and patch level changes.

You must reset the patch and minor version to 0 when you change the major version.
```
